hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/CaseOperationIntegrationTest.java

```diff
@@ -0,0 +1,446 @@
+                + "else 2 \n"
+                + "end \n"
+                + "from map", SqlColumnType.TINYINT, (byte) 100);
+
+        checkValue0("select \n"
+                + "case 1 \n"
+                + "when 1 then \n"
+                + "       case when 2 = 2 then 100 end \n"
+                + "else 2 \n"
+                + "end \n"
+                + "from map", SqlColumnType.TINYINT, (byte) 100);
+
+        checkValue0("select \n"
+                + "case 1 \n"
+                + "when 100 then 1 \n"
+                + "else \n"
+                + "       case when 2 = 2 then 100 end \n"
+                + "end \n"
+                + "from map", SqlColumnType.TINYINT, (byte) 100);
+
+        checkValue0("select t.* from \n"
+                + "(select case 1 when 1 then 100 end from map) as t", SqlColumnType.TINYINT, (byte) 100);
+
+        checkValue0("select this from map \n"
+                + "where 100 = case this when 1 then 100 end", SqlColumnType.INTEGER, 1);
+    }
+
+    @Test
+    public void testOnlyElse() {
+        put(1);
+
+        String sql = "select case else 2 end from map";
+
+        checkFailure0(sql, SqlErrorCode.PARSING, "Encountered \"case else\"");
+    }
+
+    @Test
+    public void useMapValue() {
+        put(1);
+
+        checkValue0("select case when this = 1 then 10 end from map", SqlColumnType.TINYINT, (byte) 10);
+    }
+
+    @Test
+    public void multipleConditions() {
+        put(1);
+
+        String sql = "select case when this > 1 then 10 when this = 1 then 100 end from map";
+
+        checkValue0(sql, SqlColumnType.TINYINT, (byte) 100);
+    }
+
+    @Test
+    public void doesntMatchWhen_andNoElseBranch() {
+        put(1);
+
+        String sql = "select case when this > 1 then 10 end from map";
+
+        checkValue0(sql, SqlColumnType.TINYINT, null);
+    }
+
+    @Test
+    public void differentCoercibleReturnTypes() {
+        put(1);
+
+        checkValue0(
+                "select case this \n"
+                        + "when 1 then " + Byte.MAX_VALUE + " \n"
+                        + "when 2 then " + Short.MAX_VALUE + " \n"
+                        + "when 3 then " + Integer.MAX_VALUE + " \n"
+                        + "else " + Long.MAX_VALUE + " \n"
+                        + "end from map",
+                SqlColumnType.BIGINT,
+                (long) Byte.MAX_VALUE);
+
+        LocalDateTime dateTime = LocalDateTime.of(2021, 1, 1, 10, 0, 0);
+        checkValue0(
+                "select case this \n"
+                        + "when 1 then CAST('2021-01-01T10:00' AS TIMESTAMP) \n"
+                        + "when 2 then CAST('2021-01-01' AS DATE) \n"
+                        + "else CAST('2021-01-01T10:00+00:00' as TIMESTAMP WITH LOCAL TIME ZONE) \n"
+                        + "end from map",
+                SqlColumnType.TIMESTAMP_WITH_TIME_ZONE,
+                OffsetDateTime.of(dateTime, ZoneId.systemDefault().getRules().getOffset(dateTime)));
```

Comment:
Why not use a fixed time zone?

Reply:
I couldn't find how to cast `'2021-01-01T10:00+00:00'` as timestamp with fixed time zone.
if we using `LOCAL TIME ZONE` then it will be different for CI and dev machines.

Reply:
You need to use `CAST xxx AS TIMESTAMP_WITH_TIME_ZONE`.

I tried it though, but the issue is elsewhere: the result comes from `WHEN 1` and there's a timestamp without time zone. Because the types are mixed, it is converted to TIMESTAMP WITH TIME ZONE and the system time zone is used...